### PR TITLE
Improvements to the plugin system with examples

### PR DESCRIPTION
**Pull request by ivan-cukic**

### CODE REVIEW

File: test/qa_hier_node.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        return graph;
+    }
+
+public:
+    hier_node() : _scheduler(make_graph()){};
+
+    ~hier_node() override = default;
+
+    std::string_view
+    name() const override {
```
Comment (RalphSteinhagen) [inline test/qa_hier_node.cpp:64]: maybe use node<T>::set_name(...)

File: test/qa_hier_node.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    template<fair::meta::t_or_simd<T> V>
+    [[nodiscard]] constexpr auto
+    process_one(V a, V b) const noexcept {
+        return a + b;
+    }
+};
+
+template<typename T>
+class hier_node : public fg::node_model {
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply test/qa_hier_node.cpp:34]: regarding the 'hier_node` the (at least two) options are: hier-block ...
  a) ... is a sub-graph in the sense that it merely groups blocks that are handled by the same superordinate scheduler, 
  b) ... is a sub-graph with it's own-scheduler and policies ... here the sub-graph-scheduler would need to be more like a block/node implementing a node, ... callback functions (<-> needed e.g. for GPU-domain sub-graphs)
  c) ...?

**Review by RalphSteinhagen — COMMENTED** [review]
:+1:  Looks good! 

Some more structural comments on the dynamic ports and hire functions. The main things that should IMO be addressed in this PR:
  * finding a good global naming and renaming the various uses of `tag_t::map_type`, `fair::graph::node_construction_param`, ... definitions to that.
  * The choice of having ports that are statically and/or dynamically defined should be dependent on the base class but IMO, always use the `node<T>` CTRP-base class (N.B. to be renamed -> `block<T>` to be consistent with existing GR3 naming conventions)
  * the hier_node (I guess soon named `hier_block` or perhaps `sub-graph` ) concept is good and one of the two possible/needed implementations ... we need to look for a nice API where users can declare the sub-graph in the sense of 'grouping' and or sub-domain with its own scheduler.
  
@ivan-cukic  what do you think?

File: test/qa_dynamic_node.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                available_samples);
+
+        for (auto &input_port : _input_ports) {
+            assert(available_samples == input_port.streamReader().consume(available_samples));
+        }
+        return fg::work_return_t::OK;
+    }
+
+    virtual void *
+    raw() override {
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply test/qa_dynamic_node.cpp:70]: `raw` exists in order for graph to be able to find nodes (a node doesn't know it is wrapped in something, when it wants to communicate with graph, the graph needs to be able to 'see through' the wrappers.

File: test/qa_dynamic_node.cpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+#include <list>
+
+namespace fg = fair::graph;
+
+template<typename T>
+class multi_adder : public fg::node_model {
+protected:
+    using in_port_t = fg::IN<T>;
+    // std::list because ports don't like to change in-memory address
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply test/qa_dynamic_node.cpp:11]: Opened https://github.com/fair-acc/opendigitizer/issues/90

File: test/qa_dynamic_node.cpp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    // after connection is established, and vector might reallocate
+    std::list<in_port_t> _input_ports;
+    fg::OUT<T>           _output_port;
+
+public:
+    multi_adder(std::size_t input_ports_size) {
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply test/qa_dynamic_node.cpp:17]: For later, added a comment - it needs dynamic ports to be supported by fg::node

File: test/qa_dynamic_node.cpp [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        for (auto &input_port : _input_ports) {
+            _dynamic_input_ports.emplace_back(input_port);
+        }
+        _dynamic_output_ports.emplace_back(_output_port);
+        _dynamic_ports_loaded = true;
+    };
+
+    ~multi_adder() override = default;
+
+    std::string_view
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply test/qa_dynamic_node.cpp:28]: For later, added a comment - it needs dynamic ports to be supported by fg::node

File: test/qa_dynamic_node.cpp [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        data[0]      = value;
+        data.publish(1_UZ);
+
+        value += 1;
+        return fg::work_return_t::OK;
+    }
+};
+
+template<typename T>
+class cout_sink : public fg::node<cout_sink<T>, fg::IN<T, 0, 1024, "in">> {
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply test/qa_dynamic_node.cpp:103]: https://github.com/fair-acc/opendigitizer/issues/91

**Review by RalphSteinhagen — APPROVED** [review]
LGTM :+1: !

The things that need to be tackled in later PR are documented.

**Pull request merged**